services/placement/algo/mirrored.go

```diff
@@ -0,0 +1,299 @@
+}
+
+func (a mirroredAlgorithm) ReplaceInstance(
+	p services.Placement,
+	instanceID string,
+	addingInstances []services.PlacementInstance,
+) (services.Placement, error) {
+	if !p.IsMirrored() {
+		return nil, errMirrorAlgoOnNotMirroredPlacement
+	}
+
+	if len(addingInstances) != 1 {
+		return nil, fmt.Errorf("invalid number of instances: %d for mirrored replace", len(addingInstances))
+	}
+	addingInstance := placement.CloneInstance(addingInstances[0])
+
+	// Clean up shard states before the replacement operation.
+	mirror, err := mirrorFromPlacement(p)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err = placementFromMirror(mirror, nonLeavingInstances(p.Instances()), p.ReplicaFactor())
+	if err != nil {
+		return nil, err
+	}
+
+	leavingInstance, ok := p.Instance(instanceID)
+	if !ok {
+		return nil, fmt.Errorf("instance %s not found in placement", instanceID)
+	}
+
+	shards := leavingInstance.Shards()
+	for _, s := range shards.All() {
+		// NB(cw): Mark Available as Initializing in case we use shard state
+		// to indicate cutover/cutoff time for the shard.
+		if s.State() == shard.Available {
+			s.SetState(shard.Initializing)
+		}
+	}
+
+	addingInstance = addingInstance.SetShards(shards).SetShardSetID(leavingInstance.ShardSetID())
+
+	// NB(cw): Replace is supposed to be used when the leaving node is considered dead,
+	// so the leaving node will be removed from the placement in this algo, unlike the
+	// replacement of non-mirrored algo which will leave the leaving node in the
+	// placement with all shards marked as Leaving.
+	// This is because we use shardSetID as the sourceID in mirrored placement for
+	// initializing shards, and I don't want the adding instance to keep a sourceID
+	// of the leaving instance for this case.
+	instances := placement.RemoveInstanceFromList(p.Instances(), instanceID)
+	instances = append(instances, addingInstance)
+	return placement.NewPlacement().
+		SetInstances(instances).
+		SetReplicaFactor(p.ReplicaFactor()).
+		SetShards(p.Shards()).
+		SetIsSharded(p.IsSharded()).
+		SetIsMirrored(p.IsMirrored()), nil
+}
+
+func groupInstancesByShardSetID(
+	instances []services.PlacementInstance,
+	rf int,
+) ([]services.PlacementInstance, error) {
+	var (
+		shardSetMap   = make(map[string]uint32, len(instances))
+		shardSetCount = make(map[string]int, len(instances))
+		res           = make([]services.PlacementInstance, 0, len(instances))
+	)
+	for _, instance := range instances {
+		ssID := instance.ShardSetID()
+		weight := instance.Weight()
+		shardSetCount[ssID] = shardSetCount[ssID] + 1
+		if w, ok := shardSetMap[ssID]; ok {
+			if weight == w {
+				continue
+			}
+			return nil, fmt.Errorf("found different weights: %d and %d, for shardset id %s", w, weight, ssID)
+		}
+
+		shardSetMap[ssID] = instance.Weight()
+		res = append(
+			res,
+			placement.NewInstance().
+				SetID(ssID).
+				SetRack(ssID).
```

Comment:
I'd prefer to put that logic outside of the algo because it involves instance filtering, for example if you pass in 15 instances to be added to a placement with rf=2, you can at most make 7 pairs of instances and 1 instance will be left out, I prefer to make the algo to be able to take in all instances passed in and return error otherwise. So it's better for the placement service to decide which 14 instances to make the 7 pairs of instances, and while it's doing that, assigning shardSetID to each pair is pretty easy.